File: WebServer/request-session.fs
module RequestSession
open RequestTypes
open Request
open System
open System.IO
open System.Net.Sockets
open System.Security.Authentication
open System.Text
open System.Net.Security

let private close session fullClose =
    if fullClose then
        session.networkStream.Close ()
    else
        session.tcpClient.Client.Shutdown (SocketShutdown.Send)

let private checkHeaders buffer = 
    let index = seq {0..buffer.currentIndex} |> Seq.tryFindIndex (fun i ->
        i > 4 && buffer.buffer.[i] = byte '\n' && buffer.buffer.[i - 1] = byte '\r' && buffer.buffer.[i - 2] = byte '\n')

    match index.IsSome with
    | true ->             
        let resultBuffer = {
            buffer with
                currentIndex = index.Value + 1
        } 
        {
            header = Some (Encoding.ASCII.GetString (buffer.buffer, 0, index.Value - 1))
            buffer = resultBuffer
        } 
    | false ->
        {
            header = None
            buffer = buffer
        } 

let private startReadBuffer buffer action =
    async {
        try 
            let! read = buffer.session.networkStream.AsyncRead (buffer.buffer, buffer.currentIndex, buffer.buffer.Length - buffer.currentIndex)
            if read <> 0 then
                let buffer = {
                    buffer with
                        currentIndex = buffer.currentIndex + read
                        read = read
                }
                action buffer
            else
                // TODO: Here log socket closed
                ()
        with 
            | :? SocketException as se when se.SocketErrorCode = SocketError.TimedOut -> 
                printfn "Socket session closed, Timeout has occurred"
                close buffer.session true
            | :? SocketException -> 
                close buffer.session true
            | :? ObjectDisposedException ->
                close buffer.session true
            | :? IOException ->
                close buffer.session true
            | :? AuthenticationException as ae -> 
                printfn "An authentication error has occurred while reading socket, endpoint: %s, error: %s" 
                    (buffer.session.tcpClient.Client.RemoteEndPoint.ToString()) (ae.ToString ()) 
            | ex -> 
                printfn "Socket session closed, an error has occurred while receiving: %s" (ex.ToString ())
                close buffer.session true
    } |> Async.StartImmediate

let private startReceive session configuration redirectTls =
    let buffer = {
        session = session
        buffer = Array.zeroCreate 20000
        currentIndex = 0
        read = 0
    } 
    startReadBuffer buffer <|fun buffer -> 
        let result = checkHeaders buffer 
        if result.header.IsSome then
            startRequesting result configuration session result.buffer redirectTls
        else
            startReadBuffer result.buffer |> ignore

let create (tcpClient: TcpClient) configuration certificate =
    let getNetworkStream () =
        match certificate with
        | Some certificate -> 
            let tcpStream = tcpClient.GetStream ()
            let tlsStream = new SslStream (tcpStream)
            tlsStream.AuthenticateAsServer (certificate, false, SslProtocols.Tls12 ||| SslProtocols.Tls13, false)
            tlsStream :> Stream
        | None -> tcpClient.GetStream () :> Stream
    let session = {
        tcpClient = tcpClient
        networkStream = getNetworkStream () 
        startReceive = startReceive
    }
    startReceive session configuration (configuration.UseLetsEncrypt && certificate.IsNone)

File: WebServer/RequestSession.cs
using System;
using System.Diagnostics;
using System.IO;
using System.Net;
using System.Net.Sockets;
using System.Threading;
using System.Threading.Tasks;

namespace UwebServer
{
    class RequestSession : IRequest
    {
        public DateTime? RequestStartTime { get; private set; }
        public DateTime RequestStart { get => RequestStartTime ?? DateTime.Now; }
        public IRequestHeaders Headers { get; private set; } = new RequestHeaders();
        public bool IsSecureConnection { get { return SocketSession?.UseTls ?? false; } }
		public IPEndPoint LocalEndPoint { get; private set; }
		public IPEndPoint RemoteEndPoint { get; private set; }
        public long BytesSent { get; private set; }

        public string UrlRoot
        {
            get
            {
                if (_UrlRoot == null)
                    _UrlRoot = $"http{(Server.Settings.IsTlsEnabled ? "s" : null)}://{Headers.Host}";
                return _UrlRoot;
            }
        }
        string _UrlRoot;

        public string Id { get; }

        public SocketSession SocketSession { get; private set; }

        public Server Server { get; private set; }
    
        public string HttpResponseString
        {
            get
            {
                if (_HttpResponseString == null && Headers is RequestHeaders requestHeaders)
                    _HttpResponseString = requestHeaders.Http10 ? "HTTP/1.0" : "HTTP/1.1";
                return _HttpResponseString ?? "";
            }
        }
        string _HttpResponseString;

        public RequestSession(Server server, SocketSession socketSession, Stream networkStream, DateTime? startTime)
        {
            RequestStartTime = startTime;
            Id = socketSession.Id + "-" + Interlocked.Increment(ref sessionIDCreator);
            responseHeaders = new ServerResponseHeaders(server, Id);
            stopwatch = new Stopwatch();

            SocketSession = socketSession;
            Server = server;
            this.networkStream = networkStream;

			LocalEndPoint = (SocketSession.Client.Client.LocalEndPoint as IPEndPoint)!;
			RemoteEndPoint = (SocketSession.Client.Client.RemoteEndPoint as IPEndPoint)!;

            Console.WriteLine($"{Id} New request session created - {RemoteEndPoint}");
        }

        /// <summary>
        /// 
        /// </summary>
        /// <returns>True: Continue in this SocketSession, otherwise stops processing</returns>
        public async Task<bool> StartAsync()
        {
            try
            {
                var cts = new CancellationTokenSource(Server.Settings.SocketLifetime);
                var read = await networkStream.ReadAsync(readBuffer, 0, readBuffer.Length, cts.Token);
                if (read == 0)
                {
                    Console.WriteLine($"{Id} Socket session closed");
                    return false;
                }
                if (!RequestStartTime.HasValue)
                    RequestStartTime = DateTime.Now;
                return await ReceiveAsync(read);
            }
            catch (OperationCanceledException)
            {
                Console.WriteLine($"{Id} Closing socket session, lifetime exceeded");
                Close(true);
                return false;
            }
            catch (Exception e) when (e is IOException || e is ConnectionClosedException || e is SocketException)
            {
                Console.WriteLine($"{Id} Closing socket session: {e}");
                Close(true);
                return false;
            }
            catch (Exception e) when (e is ObjectDisposedException)
            {
                Console.WriteLine($"{Id} Object disposed");
                Close(true);
                return false;
            }
            catch (Exception e)
            {
                Console.Error.WriteLine($"{Id} An error has occurred while reading socket: {e}");
                Close(true);
                return false;
            }
        }

        public void Close(bool fullClose)
        {
            try
            {
                if (fullClose)
                {
                    networkStream.Close();
                    isClosed = true;
                }
                else
                    SocketSession.Client.Client.Shutdown(SocketShutdown.Send);
            }
            catch { }
        }

        public async Task WriteAsync(byte[] buffer, int offset, int length)
        {
            await networkStream.WriteAsync(buffer, offset, length);
            BytesSent += length;
        }

        public async Task ReadStreamAsync(Stream stream)
        {
            var cls = Headers["content-length"];
            var length = long.Parse(cls ?? "0");

            while (length > 0)
            {
                int read;
                if (readFromBuffer)
                {
                    var cache = bufferReadCount - bufferEndPosition;
                    if (cache > 0)
                        read = (int)Math.Min(length, (long)cache);
                    else
                    {
                        readFromBuffer = false;
                        continue;
                    }
                }
                else
                {
                    var readLength = (int)Math.Min((long)readBuffer.Length, length);
                    read = await networkStream.ReadAsync(readBuffer, 0, readLength);
                    if (read == 0 && readLength > 0)
                        throw new ConnectionClosedException();
                }
                length -= read;
                await stream.WriteAsync(readBuffer, readFromBuffer ? bufferEndPosition : 0, read);
                readFromBuffer = false;
            }
        }

        async Task<bool> ReceiveAsync(int bufferPosition)
        {
            try
            {
                var result = await (Headers as Headers)!.InitializeAsync(networkStream, readBuffer, bufferPosition, false /*Server.Settings.HeaderTracing */, Id);
                bufferEndPosition = result.BufferEndPosition;
                readFromBuffer = bufferEndPosition > 0;
                bufferReadCount = result.BufferReadCount;

                Console.WriteLine($"{Id} Request: {RemoteEndPoint} \"{Headers.Method} {Headers.Url} {Headers.Http}\"");

                stopwatch?.Start();

				//string refererPath = null;
				//if (headers.ContainsKey("Referer"))
				//{
				//    string path = headers["Referer"];
				//    path = path.Substring(path.IndexOf("//") + 2);
				//    refererPath = path.Substring(path.IndexOf('/') + 1);
				//}

				// if (Headers.Method == Method.OPTIONS)
                //     return ServeOptions();

                return await ProcessRoutes();

                async Task<bool> ProcessRoutes()
                {
                    var response = new Response(this, responseHeaders);
                    foreach (var route in Server.Settings.Routes)
                    {
                        if (route.Method != Method.UNDEFINED && route.Method != Headers.Method)
                            continue;
                        if (route.Path != null && !Headers.Url.StartsWith(route.Path, true, null))
                            continue;
                        if (route.BasicAuthentication != null)
                        {
                            if (! await route.BasicAuthentication.Authenticate(response, Headers))
                                return false;
                        }
                        if (!await route.ProcessAsync(this, Headers, response))
                            continue;
                        return true;
                    }
                    await response.SendNotFoundAsync();
                    return false;
                }
            }
            catch (SocketException se)
            {
                if (se.SocketErrorCode == SocketError.TimedOut)
                {
                    Console.Error.WriteLine($"{Id} Socket session closed, Timeout has occurred");
                    Close(true);
                    return false;
                }
                return true;
            }
            catch (ConnectionClosedException)
            {
                Console.WriteLine($"{Id} Socket session closed via exception");
                Close(true);
                return false;
            }
			catch (ObjectDisposedException oe)
            {
                Console.WriteLine($"{Id} Socket session closed, an error has occurred: {oe}");
                Close(true);
                return false;
            }
            catch (Exception e)
            {
                Console.Error.WriteLine($"{Id} Socket session closed, an error has occurred while receiving: {e}");
                Close(true);
                return false;
            }
            finally
            {
                var elapsed = stopwatch?.Elapsed;
                stopwatch?.Stop();
                Console.WriteLine($"{Id} Answer: {RemoteEndPoint} \"{Headers.Method} {Headers.Url} {Headers.Http}\" Status: {responseHeaders.Status} Size: {responseHeaders.ContentLength} Duration: {elapsed}");
            }
        }

        static int sessionIDCreator;
        readonly object locker = new object();

        Stream networkStream;
        // has to be short living!!
        byte[] readBuffer = new byte[80000]; // Has to be under 85kB, otherwise the buffer gets to LOH
        
        ServerResponseHeaders responseHeaders;

        /// <summary>
        /// Position in buffer at which header data ends and payload starts, immediately after reading header from network stream
        /// </summary>
        int bufferEndPosition;
        /// <summary>
        /// If payload is read and available in buffer
        /// </summary>
        bool readFromBuffer;
        /// <summary>
        /// Number of read bytes already in buffer, immediately after reading header from network stream
        /// </summary>
        int bufferReadCount;
        bool isClosed;
        readonly Stopwatch stopwatch;
    }
}